Gentoo Support

JoseJX opened the pull request:
This is a bit a rough so far, but I'm interested in adding Gentoo support to Crouton. I don't want to suggest that this should be integrated as is, but I'd like some feedback and was wondering if others would be interested in this.

This currently builds most of the targets, but I haven't yet worked on the KDE, Gnome or Kodi/XBMC targets. I'm currently using a Pixel 2 for development.

Thanks!

Commit:
Gentoo: Add initial support for the core target.

Commit:
Gentoo: Add support for the audio target.

Commit:
Gentoo: Make portage rebuild packages if the use flags have changed.

Commit:
Gentoo: Add support for the cli-extra target.

Commit:
Gentoo: Add support for the xorg and x11-common targets.

Commit:
Gentoo: Fix sudo duplication and group GID.

Commit:
Gentoo: Support the gtk-extra target.

Commit:
Gentoo: Update the portage tree during a core update.

Commit:
Gentoo: Add support for the xfce and xfce-desktop targets.

Commit:
Gentoo: Add support for extension target.

Commit:
Gentoo: Don't call install when there's nothing passed to install.

Commit:
Gentoo: Add support for enlightenment.

Commit:
Gentoo: Add support for the chrome targets.

Commit:
Gentoo: Add the ability to keyword packages and accept licenses.

Commit:
Gentoo: Add support for lxde target.

Commit:
Gentoo: Add support for the xiwi target.

Commit:
Gentoo: Remove the suid bit to allow LD_PRELOAD for libcroutonfreon.

DennisLfromGA commented:
Looking forward to this, thanx.

I tried to install it but got stuck near the end I think.  

This is the command string I used to install it:

`sudo sh ./installer/main.sh -p /var/crouton/ -r gentoo -t xiwi,xorg,cli-extra,keyboard,lxde-desktop,chrome-beta`

Here's where the output ended:
```
The following USE changes are necessary to proceed:
 (see "package.use" in the portage(5) man page for more details)
# required by x11-drivers/xf86-video-intel-2.99.917_p20170216::gentoo
# required by xf86-video-intel (argument)
>=x11-libs/libdrm-2.4.75 video_cards_intel

Use --autounmask-write to write changes to config files (honoring
CONFIG_PROTECT). Carefully examine the list of proposed changes,
paying special attention to mask or keyword changes that may expose
experimental or unstable packages.
Failed to complete chroot setup.
Unmounting /var/crouton/chroots/gentoo...
```
I haven't tried to fix it yet to see what you think.

-**DennisL**

JoseJX commented:
Thanks for testing! I was adding targets one by one and didn't notice the conflict between xiwi and xorg. I'm working on a fix for that now.

JoseJX commented:
I've added a bit more intelligence to the expand variables, which should fix this issue. I've also made a few tweaks to the xfce-desktop and lxde-desktop targets to more closely match what you're expecting from debian/ubuntu, although these are completely untested so far since I haven't had time to compile them yet.

Thanks for taking a look!

DennisLfromGA commented:
@JoseJX,

> I've added a bit more intelligence to the expand variables, which should fix this issue.

I may have been a bit premature but I just updated the repo and tried an update (via '-u') but am still getting the video card issue -
```
The following USE changes are necessary to proceed:
 (see "package.use" in the portage(5) man page for more details)
# required by x11-drivers/xf86-video-intel-2.99.917_p20170216::gentoo
# required by xf86-video-intel (argument)
>=x11-libs/libdrm-2.4.75 video_cards_intel
```
I'll try to be more patient and wait until you have a chance to work on things... ;-)

-**DennisL**

JoseJX commented:
No worries! I kicked off a new build with the exact same targets as you're using above, hopefully that will help me figure out what's going on for you! One difference is that I'm building the crouton executable and installing with that, but I'm not sure how that would change the result. 

What Chrome device are you using?

You might want to try removing the VIDEO_CARDS and INPUT_DEVICES lines from /etc/portage/make.conf, it might help get over this bump!

Can you attach (from the chroot) /etc/portage/make.conf if it fails again?

DennisLfromGA commented:
@JoseJX,

> One difference is that I'm building the crouton executable and installing with that, but I'm not sure how that would change the result.

Actually, I do the same thing, I 'make' the crouton executable and install or update using it.
I don't *think* it should make any difference using 'main.sh' though.

> What Chrome device are you using?

I've got the Dell Chromebook 13 (7310) 'lulu' model (i5 / 8GB RAM / 256GB SSD) with ARC++ enabled and using -
```
Version 57.0.2987.115 beta (64-bit)
Platform 9202.54.0 (Official Build) beta-channel lulu
ARC Version 3828404
Firmware Google_Lulu.6301.136.57
```
> You might want to try removing the VIDEO_CARDS and INPUT_DEVICES lines from /etc/portage/make.conf, it might help get over this bump!

I'll do that now and get back to you.

> Can you attach (from the chroot) /etc/portage/make.conf if it fails again?

Will do.

Thanx,
-**DennisL**

Commit:
Gentoo: Support use expand variables for simultaneous xiwi and xorg.

DennisLfromGA commented:
@JoseJX,

Just an update, I know you're working on it now but I thought I'd let you know what I ran into.
Here's my `make.conf`
```
chronos@localhost /var/crouton/chroots/gentoo/etc/portage $ cat make.conf
# These settings were set by the catalyst build script that automatically
# built this stage.
# Please consult /usr/share/portage/config/make.conf.example for a more
# detailed example.
CFLAGS="-O2 -pipe -march=native -mtune=native"
CXXFLAGS="${CFLAGS}"
# WARNING: Changing your CHOST is not something that should be done lightly.
# Please consult http://www.gentoo.org/doc/en/change-chost.xml before changing.
CHOST="x86_64-pc-linux-gnu"
# These are the USE and USE_EXPAND flags that were used for
# buidling in addition to what is provided by the profile.
USE="bindist pulseaudio alsa X dbus xcb xkb libkms uxa dri3 xvmc opengl"
CPU_FLAGS_X86="aes avx avx2 fma3 mmx mmxext popcnt sse sse2 sse3 sse4_1 sse4_2 ssse3"
PORTDIR="/usr/portage"
DISTDIR="${PORTDIR}/distfiles"
PKGDIR="${PORTDIR}/packages"
ABI_X86="32 64"
MAKEOPTS="-j5"
##VIDEO_CARDS="dummy"
##INPUT_DEVICES="keyboard libinput mouse evdev"
```
And here are the errors I'm running into when I enter the chroot and try to let it finish the setup.
```
The following USE changes are necessary to proceed:
 (see "package.use" in the portage(5) man page for more details)
# required by x11-drivers/xf86-video-ati-7.8.0::gentoo[glamor]
# required by x11-base/xorg-drivers-1.19::gentoo[video_cards_radeonsi]
>=x11-base/xorg-server-1.19.2 glamor

Use --autounmask-write to write changes to config files (honoring
CONFIG_PROTECT). Carefully examine the list of proposed changes,
paying special attention to mask or keyword changes that may expose
experimental or unstable packages.

emerge: there are no ebuilds built with USE flags to satisfy "x11-base/xorg-server[glamor]".
!!! One of the following packages is required to complete your request:
- x11-base/xorg-server-1.18.4::gentoo (Change USE: +glamor)
(dependency required by "x11-drivers/xf86-video-ati-7.8.0::gentoo[glamor]" [ebuild])
(dependency required by "x11-base/xorg-drivers-1.19::gentoo[video_cards_radeonsi]" [ebuild])
(dependency required by "x11-base/xorg-server-1.19.2::gentoo[xorg]" [ebuild])
(dependency required by "xorg-server" [argument])
Failed to complete chroot setup.
The chroot setup script may be broken. Your chroot is not fully configured.
Removing the chroot setup script. You may want to update your chroot again.
UID 1000 not found in gentoo
Unmounting /var/crouton/chroots/gentoo...
Sending SIGTERM to processes under /var/crouton/chroots/gentoo...
```
Let me know if you'd like me to try anything else, otherwise I'll just hold off for a while.

Thanx,
-**DennisL**

JoseJX commented:
Okay, so I hadn't thought of the variables being commented out like that! I've changed it so that the match should check for the variable at the start of the line which I think should fix that issue. I'm currently away from my chromebook, so I'm not sure yet if my build that I started earlier was successful, I'll follow up on that later tonight. 

The errors following your make.conf are from there not being a valid VIDEO_CARDS line set, which just defaults to a normally sane set of cards, but those require a few more use flags to be set.

dnschneid commented:
This is fantastic!  I'll review it in multiple passes (especially since you're still working on it), but at a quick glance it looks like a really good start.

dnschneid commented on installer/gentoo/bootstrap:
```diff
@@ -0,0 +1,73 @@
+# from $RELEASE), $ARCH, and $MIRROR.
+
+# Set the arch if it's not already set
+if [ -z $ARCH ]; then
```
ARCH is already set, so shouldn't be set here.

dnschneid replied:
(oops, pressed the wrong button...starting a review)

dnschneid commented on installer/gentoo/defaults:
```diff
@@ -0,0 +1,19 @@
+*) error 2 "Invalid architecture '$ARCH'.";;
+esac
+
+BOOTSTRAP_RELEASE="gentoo"
```
This isn't necessary.

dnschneid commented on installer/gentoo/getrelease.sh:
```diff
@@ -0,0 +1,27 @@
+if [ -f "${2%/}/etc/gentoo-release" ]; then
+    # Print the architecture if requested
+    if [ "$1" = '-a' ]; then
+        echo `readlink "${2%/}/etc/portage/make.profile" | cut -d"/" -f8`
```
This seems unlikely to work if crouton is installed into something other than /usr/local/chroots (@DennisLfromGA is/will probably hit this).

dnschneid replied:
This may not give a good output once the other arches are supported.

dnschneid replied:
```echo `...` ``` is probably unnecessary, since you can just let the command output directly.  Of course, once you integrate the other arches, you may need to parse the output of this to produce the correct result.

dnschneid replied:
Is there a better way to check the architecture, other than checking a symlink?

dnschneid commented on installer/gentoo/getrelease.sh:
```diff
@@ -0,0 +1,27 @@
+fi
+
+# Check if this is gentoo
+if [ -f "${2%/}/etc/gentoo-release" ]; then
```
If you invert this, you can exit 1 early, and then the rest of the script won't be encased in an if statement.

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,179 @@
+# Set global use flags, separated by a space
+set_flags() {
+    while [ "$#" != 0 ]; do
+         local flag=`euse -a $1 | cut -d\[ -f1`;
```
There's an extra space here

dnschneid replied:
What's the output you are parsing here?  There might be a cleaner way to do this.

dnschneid commented on installer/gentoo/bootstrap:
```diff
@@ -0,0 +1,73 @@
+# Fetch the latest portage tree
+p="http://distfiles.gentoo.org/snapshots/portage-latest.tar.bz2"
+pf="portage-latest.tar.bz2"
+mkdir "${DIR}/usr/portage"
```
Probably should use `-p`

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,179 @@
+# Usage: set_package_flag <package> <use flag> <crouton target for comment>
+set_package_flag() {
+    # Check if the directory exists
+    if [ ! -d /etc/portage/package.use ]; then
```
You can unconditionally mkdir if you use `-p`

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,179 @@
+    fi
+
+    # Check if the package use file exists
+    if [ ! -f /etc/portage/package.use/crouton ]; then
```
It's fine to touch existing files.

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,179 @@
+    fi
+
+    # Check if the requested package/use is already in the file
+    USEFLAG=`grep -e "^$1 $2$" /etc/portage/package.use/crouton | cut -f1 -d1`
```
Maybe `grep -F` would be better

dnschneid replied:
I'm not a fan of the `| cut ...; if [ -z` template.  Can you use the return value of grep?

dnschneid replied:
That is, `-F` and `-x`

dnschneid replied:
(here and elsewhere)

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,179 @@
+    fi
+
+    # Check if the requested package is already in the file
+    UNMASK=`grep -e "^$1[[:space:]]" -e "^$1$" /etc/portage/package.keywords/crouton | cut -f1 -d1`
```
You explicitly put spaces in, so why do you need to use `[[:space:]]`?

dnschneid replied:
You can also use `-w`

dnschneid replied:
Oh, I see, this matches how you access package.license (which I assume can be edited by other tools)

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,179 @@
+    fi
+
+    # Let the user know that we accepted the license for them
+    TIPS="$TIPS
```
Nice!!

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,179 @@
+            shift
+            break
+        fi
+	# We ignore --minimal and --asdeps it does nothing for gentoo
```
be careful with indentation (here and elsewhere)

dnschneid replied:
still relevant

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,179 @@
+
+# install_pkg_dist: see install_pkg() in prepare.sh for details.
+install_pkg_dist() {
+	echo "FIXME: install_pkg_dist"
```
maybe these should `error 99`?

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,179 @@
+
+
+# list_uninstalled_dist: see list_uninstalled() in prepare.sh for details.
+# If the package is virtual (e.g. libc-dev), we need to find the binary package
```
These comments probably don't apply to gentoo :P

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,179 @@
+}
+
+
+# install_mirror_package [--asdeps] package path [regex arch]
```
This is an ubuntu-specific thing that you probably don't need in here.

dnschneid commented on installer/prepare.sh:
```diff
@@ -198,9 +198,15 @@ compile() {
+	pkgs="gcc libc6-dev"
+    fi
+    pkgs="${pkgs} $*"
+    if ! [ -z $pkgs ]; then
```
It's fine to call install with an empty package list, iirc

dnschneid commented on installer/prepare.sh:
```diff
@@ -198,9 +198,15 @@ compile() {
         shift 1
     fi
     shift 2
+    local pkgs=""
```
Set pkgs to "$*" and prefix in the debian branch?

Alternatively, I think you could just `debian=gcc, debian=libc6-dev` and be done with it.

dnschneid commented on targets/audio:
```diff
@@ -3,15 +3,19 @@
 # found in the LICENSE file.
 REQUIRES='core'
-DESCRIPTION='Support audio playback via Chromium OS'\''s audio system.'
+DESCRIPTION='Support audio playback via Chromium OS audio system (CRAS).'
```
haha fine

JoseJX replied:
Acronyms are hard and I always forget what they actually mean. This was more of a "hey, remember this", I don't really need this to go in, haha.

dnschneid commented on targets/audio:
```diff
@@ -67,42 +71,52 @@ build_cras() {
+            ALSALIBDIR="/usr/lib$archextrapath/alsa-lib"
+        else
+            install --minimal --asdeps libasound2-dev$pkgsuffix
+            ALSALIBDIR="`PKG_CONFIG_PATH="$pkgconfigpath" \
```
Avoid duplicating this if possible

dnschneid commented on targets/cli-extra:
```diff
@@ -8,8 +8,13 @@ HOSTBIN='startcli'
 
+# Set the use flags on gentoo that we'll need
+if [ "${DISTROAKA:-"$DISTRO"}" = 'gentoo' ]; then
+    set_flags dbus
```
shouldn't this really be in core?

dnschneid commented on targets/core:
```diff
@@ -106,18 +106,56 @@ EOF
-    if [ "${DISTROAKA:-"$DISTRO"}" = 'debian' ]; then
-        dpkg-reconfigure locales
+CORE_pkgs="gentoo=app-admin/sudo,sudo wget ca-certificates"
+if [ "${DISTROAKA:-"$DISTRO"}" = 'debian' ]; then
```
can use `debian=apt-transport-https,` instead

dnschneid commented on targets/xfce-desktop:
```diff
@@ -7,8 +7,25 @@ DESCRIPTION='Installs Xfce along with common applications. (Approx. 1200MB)'
+
+if [ "${DISTROAKA:-"$DISTRO"}" = 'gentoo' ]; then
+    # Use this target to install the rest of the xfce4 packages and more
+    install orage mousepad ristretto squeeze thunar-archive-plugin \
```
No meta package at all?  This seems fragile...

JoseJX replied:
Gentoo is kind of minimal, it expects you to install the stuff you need. There are meta-packages, but they're not as encompassing as something like the *-desktop targets here.

That said, if you'd rather not have a giant list of packages here, I could make an overlay (kind of like a ppa) with Gentoo meta-package ebuilds that do the same thing.

dnschneid replied:
An overlay isn't any better...I'm just worried that among a long list of packages, one may go missing.  But if the alternative is to place that long list elsewhere, it's preferable to keep it here.

dnschneid commented on targets/e17:
```diff
@@ -11,7 +11,10 @@ HOSTBIN='starte17'
 ### Append to prepare.sh:
-install e17
+if [ "${DISTROAKA:-"$DISTRO"}" = 'gentoo' ]; then
+    set_flags xrandr xcomposite
```
This is really only needed in e17?

JoseJX replied:
Yes, these flags aren't used in other packages, the functionality is forced on. Gentoo is weird sometimes. :)

dnschneid commented on targets/extension:
```diff
@@ -19,7 +19,13 @@ fi
-install x11-utils xclip
+EXT_APPS=""
+if [ "${DISTROAKA:-"$DISTRO"}" = 'gentoo' ]; then
+    EXT_APPS="x11-misc/xclip appres luit xdpyinfo xdriinfo xev xfd xfontsel xkill xlsatoms xlsclients xlsfonts xmessage xprop xvinfo xwininfo"
```
most of these aren't actually needed by crouton.  Can you search through the crouton source and only include the stuff that's needed?

dnschneid replied:
no sense in setting a variable if you're just going to call install.  Just run install directly.

dnschneid commented on targets/extension:
```diff
@@ -35,12 +41,14 @@ if [ "$XMETHOD" = 'xorg' ] || grep -q xorg /etc/crouton/targets; then
+    handler='/usr/local/bin/croutonurlhandler'
+    for link in x-www-browser gnome-www-browser www-browser; do
+        if ! update-alternatives --query "$link" | grep -q "$handler"; then
+            update-alternatives --install "/usr/bin/$link" "$link" "$handler" 15
```
Is there an equivalent for alternatives in gentoo?

JoseJX replied:
I'm not really aware of one, but I'll look into it some more.

dnschneid replied:
I guess the question is, is there a common script in gentoo that launches a browser -- and if so, where is it getting its preferences from?

JoseJX replied:
Ah, I see what you're asking now. No, I'm not aware of anything that does that for Gentoo. Things like gnome/kde do that with application defaults, but there's nothing in Gentoo that does that as far as I know.

dnschneid replied:
Can you check the gnome/kde scripts and see if they fall back on something that might be common?

JoseJX replied:
Here's what the Gentoo docs say about setting default applications: https://wiki.gentoo.org/wiki/DefaultApplication

To the best of my knowledge, packages will install freedesktop compatible MIME entries for default applications, so things like Chrome should "just work" providing that the DE supports it. I'll look into duplicating this for the crouton url handler.

dnschneid replied:
So I guess we need to add .desktop files for croutonurlhandler, which is probably a good idea anyway.  We may also want a desktop file for switching back to Chromium OS without creating a new tab (or maybe it can be the same desktop file).  Either way, I'll file a feature request; I don't think you need to include that in your PR unless you really want to :)

dnschneid commented on targets/gtk-extra:
```diff
@@ -7,18 +7,27 @@ DESCRIPTION='GTK-based tools including gdebi, gksu, and a simple browser.'
+    install gksu
+else
+    install_dummy network-manager network-manager-gnome
+    install gdebi gksu
```
`install debian=gdebi, gksu`?

dnschneid commented on targets/gtk-extra:
```diff
@@ -7,18 +7,27 @@ DESCRIPTION='GTK-based tools including gdebi, gksu, and a simple browser.'
-    fi
-done
+if [ "${DISTROAKA:-"$DISTRO"}" = 'gentoo' ]; then
+    install dillo
```
netsurf doesn't exist?

JoseJX replied:
It does, but it's masked as unstable. I wasn't sure why, so I simplified.

dnschneid replied:
sounds good to me

dnschneid commented on targets/lxde-desktop:
```diff
@@ -7,4 +7,9 @@ DESCRIPTION='Installs LXDE along with common applications. (Approx. 800MB)'
 ### Append to prepare.sh:
-install ubuntu=lubuntu-desktop,task-lxde-desktop -- network-manager
+if [ "${DISTROAKA:-"$DISTRO"}" = 'gentoo' ]; then
+    install lightdm lxlauncher lxtask galculator gpicview leafpad \
```
If you're going to list stuff out...lightdm is never needed in crouton

dnschneid commented on targets/x11-common:
```diff
@@ -18,17 +18,38 @@ if [ -n "$offset" ]; then
+
+# On Gentoo, we need to set the text useflag for xdg-utils
+if [ "${DISTROAKA:-"$DISTRO"}" = 'gentoo' ]; then
+    echo "app-text/xmlto text" >> /etc/portage/package.use/package.use
```
Every time you update the chroot, package.use will get longer...

dnschneid commented on targets/xiwi:
```diff
@@ -28,8 +28,27 @@ if [ "${DISTROAKA:-"$DISTRO"}" = 'debian' ]; then
+    set_use_expand INPUT_DEVICES evdev
+    
+    # Set use flags for X
+    set_flags X dbus
```
wasn't this done in x11-common?

dnschneid commented on targets/xiwi:
```diff
@@ -28,8 +28,27 @@ if [ "${DISTROAKA:-"$DISTRO"}" = 'debian' ]; then
     apt-mark unhold xserver-xorg-video-dummy$ltspackages || true 2>/dev/null
```
Is there an equivalent to apt-mark hold to prevent portage from recompiling dummy without the patches?

JoseJX replied:
No, but it might make more sense to add the patches as a user patch so that every time it's updated, it'll just automatically apply. I'll look into doing this instead for this package.

JoseJX replied:
Okay, two dumb questions on this:
1. Is there a reason why the driver has to overwrite the dummy driver? Why not install it as a different driver (crouton_dummy_drv.so perhaps?) and then change the X configuration to load that driver instead?
2.  Is there a reason why the patch is inlined in the shell script rather than sitting in a separate file? I was thinking of applying it as an overlay patch to the Gentoo package, which would also resolve the issue of the driver being overwritten.

Thanks!

dnschneid replied:
@drinkcat to add flavor

> Is there a reason why the driver has to overwrite the dummy driver? Why not install it as a different driver (crouton_dummy_drv.so perhaps?) and then change the X configuration to load that driver instead?

Probably for a couple reasons

1. It keeps the patch smaller
2. It's a strict superset of dummy, so it's compatible with anything already trying to use dummy
3. The intent is that the changes will be upstreamed, so we don't want a separate crouton dummy driver hanging around when the patch is no longer necessary

> Is there a reason why the patch is inlined in the shell script rather than sitting in a separate file? I was thinking of applying it as an overlay patch to the Gentoo package, which would also resolve the issue of the driver being overwritten.

Probably because there's no clean way to do that in crouton at the moment, and we didn't want to leave files lying around.  An overlay patch in Gentoo would make a lot of sense, so maybe in this case, add the patch to chrootetc?  Then the prepare script can run patch on that file, and you can symlink it into the local Gentoo overlay.

dnschneid reviewed (changes requested):
Don't be discouraged by the number of comments -- this is already in really good shape.

I tried to avoid repeating the style comments too many times, so please go through everything for style.  When things settle, I'll start pointing out everything 😱

JoseJX commented on installer/gentoo/bootstrap:
```diff
@@ -0,0 +1,73 @@
+fi
+
+# Set beter cflags
+sed -i 's/CFLAGS="-O2 -pipe"/CFLAGS="-O2 -pipe -march=native -mtune=native"/' ${DIR}/etc/portage/make.conf
```
Haha, I mean, I guess we could, got to perpetuate those Gentoo stereotypes! :)

JoseJX commented:
No worries on the number of comments, and thanks for reviewing! I'll look into addressing some of the issues tonight.

Commit:
Gentoo: Add support for RSYNC_PROXY.

Commit:
Gentoo: Add support for MIRROR, remove debugging code.

Commit:
Gentoo: Use CHOST from make.conf to determine architecture.

Commit:
Gentoo: Simplify prepare functions.

Commit:
Gentoo: Simplify changes to prepare.sh.

Commit:
Gentoo: Simplify packages for core target. Add dbus to global use flags.

Commit:
Gentoo: Clean up audio target changes, remove redundant check.

Commit:
Gentoo: Remove extra dbus use flag setting from cli-extra.

Commit:
Gentoo: Remove unused X utils from extension target.

Commit:
Gentoo: Simplify the gtk-extra target's install logic.

Commit:
Gentoo: Fix lxde-desktop target, remove lightdm.

Commit:
Gentoo: Fix use flags, clean up package installation for X targets.

JoseJX commented:
I think I've addressed most of the issues from the first review. I still haven't tested most of the changes yet, so please don't feel like you need to review this again immediately. I'm going to try to rerun DennisLfromGA's setup and make sure everything is working. I'll update this once it finishes compiling.

Thanks again for looking at the code!

Commit:
Gentoo: Update cpu flag code for cpuid2cflags >=2.

JoseJX commented:
Okay! This is in a bit better shape now. I was able to install xiwi and xorg simultaneously and both seem to be working. DennisLfromGA, I think that you should be able to complete your installation now, I was able to do a clean install with the targets you specified above.

Commit:
Gentoo: Add ability to turn off use flags in prepare.

Commit:
Gentoo: Make sure to trigger a world update in core.

dnschneid commented on installer/gentoo/defaults:
```diff
@@ -0,0 +1,21 @@
+    ARCH=$(uname -m)
+fi
+
+case "$ARCH" in
```
Still need the other arches

dnschneid commented on installer/gentoo/defaults:
```diff
@@ -0,0 +1,21 @@
+# It must set at least ARCH and MIRROR if not already specified.
+
+if [ -z "$ARCH" ]; then
+    ARCH=$(uname -m)
```
quoting

dnschneid replied:
(going to stop pointing these out this time around :) )

dnschneid commented on installer/gentoo/bootstrap:
```diff
@@ -0,0 +1,44 @@
+
+# Grab the latest stage3
+echo 'Finding the latest stage3...' 1>&2
+LATEST_STAGE3="$MIRROR/releases/$ARCH/autobuilds/latest-stage3-$ARCH.txt"
```
This isn't going to work when it comes to arm or x86, since the path will look more like `$MIRROR/releases/$ARCH_FAMILY/autobuilds/latest-stage3-$ARCH.txt`

dnschneid commented on installer/gentoo/bootstrap:
```diff
@@ -0,0 +1,44 @@
+LATEST_STAGE3="$MIRROR/releases/$ARCH/autobuilds/latest-stage3-$ARCH.txt"
+# Read the latest release, the sed matches the lines that don't start with # 
+# and then removes the size (everything after the space) from the filename
+STAGE3=$(wget -q -O- "$LATEST_STAGE3" | sed -n -e '/^#/! s/ .*//p')
```
quote the output of commands `"$(...)"`

dnschneid commented on installer/gentoo/bootstrap:
```diff
@@ -0,0 +1,44 @@
+mkdir -p "$DIR"
+
+echo "Downloading release: $STAGE3..." 1>&2
+if ! wget -O- --no-verbose --timeout=60 -t2 "$MIRROR/releases/$ARCH/autobuilds/$STAGE3" | tar -C "$DIR" --strip-components=1 -jx 2>/dev/null; then
```
keep lines <=80 characters, please

dnschneid replied:
here and elsewhere

dnschneid commented on installer/gentoo/bootstrap:
```diff
@@ -0,0 +1,44 @@
+    echo 'ABI_X86="32 64"' >> "$DIR/etc/portage/make.conf"
+fi
+
+# Set beter cflags -- no, not -funroll-loops :)
```
aww

dnschneid commented on installer/gentoo/bootstrap:
```diff
@@ -0,0 +1,44 @@
+sed -i 's/CFLAGS="-O2 -pipe"/CFLAGS="-O2 -pipe -march=native -mtune=native"/' "$DIR/etc/portage/make.conf"
+
+# Set better make opts
+CORE_COUNT=$(grep -c processor /proc/cpuinfo)
```
probably unnecessary, but quote anyway

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,131 @@
+set_flags() {
+    while [ "$#" != 0 ]; do
+        # Check if the flag starts with a -, if so, we're disabling it
+	if [ "${1:0:1}" = "-" ]; then
```
bashism.  Use this instead: `if [ "${1#-}" != "$1" ]; then`

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,131 @@
+	    euse -D "${1#"-"}" 1> /dev/null
+	else
+            # If we haven't already set the flag, we can enable it
+            if [ -z "$(euse -a $1)" ]; then
```
quote `$1`

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,131 @@
+    touch /etc/portage/package.use/crouton
+
+    # Set the use flag if it's not already set
+    if [ $(grep -c -F -x -e "$1 $2" /etc/portage/package.use/crouton || true) = "0" ]; then
```
If you're just checking if it's there or not, use `! grep -q`

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,131 @@
+    touch /etc/portage/package.keywords/crouton
+
+    # Check if the keyword is set for this package and if not, set it
+    if [ $(grep -c -e "^$1[[:space:]]" -e "^$1$" /etc/portage/package.keywords/crouton || true) = "0" ]; then
```
same here:

`if ! grep -q -e "^$1[[:space:]]" -e "^$1$" /etc/portage/package.keywords/crouton; then`

dnschneid replied:
here and elsewhere

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,131 @@
+keyword_package() {
+    # Make sure the keywords directory and file exist
+    mkdir -p /etc/portage/package.keywords
+    touch /etc/portage/package.keywords/crouton
```
Make a local variable so you don't have to repeat the path so many times?

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,131 @@
+            break
+        fi
+	# We ignore --minimal and --asdeps it does nothing for gentoo
+	if [ ! "$1" = "--minimal" -a ! "$1" = "--asdeps" ]; then
```
I think @drinkcat found that `!=` is totally legit and we should be using it

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,131 @@
+	fi
+        shift
+    done
+    if [ ! -z "$pkgs" ]; then
```
`! -z` -> `-n`

dnschneid commented on targets/audio:
```diff
@@ -67,30 +71,35 @@ build_cras() {
     # precise does not allow libasound2-dev and libasound2-dev:i386 to be
     # installed simultaneously
-    if release -le precise && [ -n "$cras_arch" ]; then
+    if release -le precise && [ -n "$cras_arch" ] && [ "${DISTROAKA:-"$DISTRO"}" = 'debian' ]; then
```
`release -le precise` will return false if the distro isn't ubuntu/debian, so the extra condition is unnecessary

dnschneid commented on targets/chrome-common:
```diff
@@ -37,20 +37,44 @@ if [ -n "`list_uninstalled '' "$pkg"`" ]; then
+	fi
+	
+	# Keyword Chrome
+	keyword_package $GENTOO_PKG $pkg
```
quotes

dnschneid commented on targets/core:
```diff
@@ -106,18 +106,62 @@ EOF
 # On release upgrade, keyboard-configuration might be reconfigured.
 fixkeyboardmode
 
+# 
```
empty comment?

dnschneid commented on targets/core:
```diff
@@ -106,18 +106,62 @@ EOF
+    # Install gentoolkit for euse, so we can set use flags
+    # Install cpuid2cpuflags so we can easily set reasonable CFLAGS
+    install gentoolkit cpuid2cpuflags
+    if [ "$ARCH" = 'x86' -o "$ARCH" = 'amd64' ]; then
```
this may need to be changed depending on whether you want ARCH to be x86 or i686 (same with arm -- armhf or armv7-hardf or whatever it is).

dnschneid replied:
And what happens on ARM?

dnschneid commented on targets/extension:
```diff
@@ -19,7 +19,7 @@ fi
 
 ### Append to prepare.sh:
-install x11-utils xclip
+install gentoo=x11-misc/xclip,xclip debian=x11-utils, gentoo=xdpyinfo, gentoo=xdriinfo, gentoo=xprop,
```
you could probably combine one of these with the x11-utils selector: `gentoo=xprop,x11-utils`

dnschneid commented on targets/x11-common:
```diff
@@ -19,16 +19,21 @@ if [ -n "$offset" ]; then
-compile xi2event '-lX11 -lXi' libx11-dev libxi-dev
-install --minimal dbus xdg-utils
+if [ "${DISTROAKA:-"$DISTRO"}" = 'gentoo' ]; then
+    compile xi2event '-lX11 -lXi' x11-libs/libXi
```
I think the dependencies list is a crouton package selector, so you could combine these into `gentoo=x11-libs/libXi,libxi-dev gentoo=,libx11-dev`

dnschneid commented on targets/xorg:
```diff
@@ -64,11 +87,17 @@ if [ "${ARCH#arm}" = "$ARCH" ]; then
     # Unhold xserver-xorg-video-intel, modification will be reapplied later.
-    apt-mark unhold "$inteldriver" 2>/dev/null || true
+    if [ "${DISTROAKA:-"$DISTRO"}" = 'debian' ]; then
+        apt-mark unhold "$inteldriver" 2>/dev/null || true
```
what are you going to do about the binary hack on the intel driver?  In this case, you could probably also make an overlay patch...

dnschneid reviewed (changes requested):
Vast improvements; thanks!  This is exciting...

JoseJX commented:
I've tested the additional architecture code with the x86 target and it seemed to work fine. I've dug out my old Samsung Chromebook 3 for testing the ARM targets. It's crazy how slow this machine is now. :(

I think I've addressed most of your comments from the second review, but I'll continue going over the code some more to tighten things up.

DennisLfromGA commented:
@JoseJX,

I was not able to repair my first chroot with targets: 
> `xiwi,xorg,cli-extra,keyboard,lxde-desktop,chrome-beta`

I kept getting hung up on the video driver issue -
```
The following USE changes are necessary to proceed:
 (see "package.use" in the portage(5) man page for more details)
# required by x11-drivers/xf86-video-ati-7.8.0::gentoo[glamor]
# required by x11-base/xorg-drivers-1.19::gentoo[video_cards_radeonsi]
>=x11-base/xorg-server-1.19.2 glamor
```

But I was able to successfully install & run a brand new one.  
I just let it set the default xorg xmethod and it works fine.
```
name: gentoo
encrypted: no
Entering /var/crouton/chroots/gentoo...
crouton: version 1-20170327171751
release: gentoo
architecture: amd64
xmethod: xorg
targets: cli-extra,keyboard,lxde-desktop,chrome-beta
host: version 9202.54.0 (Official Build) beta-channel lulu 
kernel: Linux localhost 3.14.0 #1 SMP PREEMPT Sat Mar 18 09:21:47 PDT 2017 x86_64 Intel(R) Core(TM) i5-5300U CPU @ 2.30GHz GenuineIntel GNU/Linux
freon: yes
```

I may get brave and try to add 'xiwi' but I'll back it up first.

Meanwhile I'll be studying the ['Gentoo Cheat Sheet'](https://wiki.gentoo.org/wiki/Gentoo_Cheat_Sheet) ;)

Thanx for all your efforts,
-**DennisL**

JoseJX commented:
Hey, that's great! I'm glad you were able to get it to work. I haven't tested a whole lot of the changes from last night since I'm not in front on my Chromebook right now, so expect some breakage, but I think it's getting closer to being actually usable!

DennisLfromGA commented:
Just an update -

It seems the switching key-combo <kbd>ctrl</kbd> + <kbd>alt</kbd> + <kbd>shift</kbd> + <kbd>left-arrow/right-arrow</kbd> gets broken. 
Once I switch out of `gentoo` I can't get back to any of my running chroots, to include `gentoo`. 
The 'crouton integration' extension lists them but when clicked nothing happens. 
Also `croutoncycle` in `gentoo` works once and then won't work again in any other chroot, it just hangs on -
```
+ force=
+ [ cros != cros ]
+ cmd=0
+ mkdir -m 775 -p /tmp/crouton-lock
+ exec
+ chmod -Rf g+rwX /tmp/crouton-lock
+ chgrp -Rf crouton /tmp/crouton-lock
+ flock 3
```
I've tried `croutoncycle [list|cros|:1|:2]` and it hangs at the same point.
So I'm kind of locked out of any running desktop gui.

-**DennisL**

JoseJX commented:
Hmm, I'll have to try that later, I've never actually tried switching on my machine. I'd suspect that it's missing an application or configuration option (perhaps a Debian default?). In any case, thanks for pointing that out!

dnschneid commented on installer/gentoo/getrelease.sh:
```diff
@@ -0,0 +1,37 @@
+    amd64 | x86_64) echo "amd64";;
+    x86 | i686 | i486) echo "x86";;
+    arm64 | aarch64) echo "arm64";;
+    arm | arm*) echo "arm";;
```
This isn't going to be the correct result.  According to your defaults file, ARCH needs to be armv7a_hardfloat or armv7a.

dnschneid commented on installer/gentoo/getrelease.sh:
```diff
@@ -0,0 +1,37 @@
+
+    case "$CHOST" in
+    amd64 | x86_64) echo "amd64";;
+    x86 | i686 | i486) echo "x86";;
```
`i?86`?

dnschneid commented on installer/gentoo/defaults:
```diff
@@ -0,0 +1,25 @@
+
+case "$ARCH" in
+amd64 | x86_64) ARCH="amd64";;
+x86 | i686) ARCH="x86";;
```
`i?86`?

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,136 @@
+}
+
+# list_uninstalled_dist: see list_uninstalled() in prepare.sh for details.
+# If the package is virtual (e.g. libc-dev), we need to find the binary package
```
this comment probably isn't relevant

dnschneid commented on targets/audio:
```diff
@@ -67,26 +71,32 @@ build_cras() {
     install --minimal alsa-utils \
-        libasound2$pkgsuffix libspeexdsp1$pkgsuffix
+        gentoo=media-libs/alsa-lib,libasound2$pkgsuffix \
+	gentoo=speex,libspeexdsp1$pkgsuffix
```
(nit) indentation

dnschneid commented on targets/chrome-common:
```diff
@@ -37,20 +37,45 @@ if [ -n "`list_uninstalled '' "$pkg"`" ]; then
+        # Rename google-chrome-stable to google chrome to match the ebuild
+        if [ "$CHANNEL" = 'stable' ]; then
+            GENTOO_PKG="www-client/google-chrome"
+	else
```
Please replace tabs with spaces

dnschneid commented on targets/core:
```diff
@@ -106,18 +106,64 @@ EOF
+if [ "${DISTROAKA:-"$DISTRO"}" = 'gentoo' ]; then
+    # Install gentoolkit for euse, so we can set use flags
+    # Install cpuid2cpuflags so we can easily set reasonable CFLAGS
+    install gentoolkit cpuid2cpuflags
```
does cpuid2cpuflags need to be installed on ARM at all, then?

dnschneid commented on targets/core:
```diff
@@ -106,18 +106,64 @@ EOF
+    install gentoolkit cpuid2cpuflags
+    # This tool isn't really supported on ARM yet, update this when it is
+    if [ "$ARCH" = 'x86' -o "$ARCH" = 'amd64' ]; then
+        CPUFLAGS=$(cpuinfo2cpuflags-x86 | cut -f2 -d' ')
```
quote the output?

dnschneid commented on targets/core:
```diff
@@ -106,18 +106,64 @@ EOF
+# Debian uses GID 28 for sudo
+if [ "${DISTROAKA:-"$DISTRO"}" = 'gentoo' ]; then
+    # If the sudo group already exists, we'll assume we already set it up
+    if [ -z "$(getent group sudo | cut -f1 -d:)" ]; then
```
`if ! grep -q ^sudo /etc/group; then` ?

dnschneid commented on targets/core:
```diff
@@ -106,18 +106,64 @@ EOF
+
+if [ "${DISTROAKA:-"$DISTRO"}" = 'debian' ]; then
+    # Generate and set default locale
+    if [ ! -f '/etc/default/locale' ] && hash locale-gen 2>/dev/null; then
```
does locale-gen really not need to happen on gentoo?

dnschneid replied:
if so, the debian check below is redundant

JoseJX replied:
locale-gen is run when glibc is installed. By default, it generates all of the possible locales. If you don't want all of the locale data, you can set /etc/locale.gen to specify what locales you'd like to actually build.

I'll change it to set the default locale to en_US.UTF-8 though, that can't hurt.

dnschneid commented on targets/xfce:
```diff
@@ -34,8 +34,16 @@ if release -lt xenial && [ "$ARCH" = 'arm64' ]; then
+    # NOTE: xfce4-notifyd is in here to avoid pulling in the gnome version
+    install xfce4-meta xfce4-notifyd xfce4-pulseaudio-plugin tango-icon-theme \
+        thunar xfce4-mixer xfce4-power-manager \
+    	xfce4-terminal xfce4-battery-plugin xfce4-xkb-plugin
```
nit: tab

dnschneid commented on targets/xfce-desktop:
```diff
@@ -7,8 +7,25 @@ DESCRIPTION='Installs Xfce along with common applications. (Approx. 1200MB)'
+    # Use this target to install the rest of the xfce4 packages and more
+    install orage mousepad ristretto squeeze thunar-archive-plugin \
+        thunar-media-tags-plugin xfburn xfce4-clipman-plugin \
+	xfce4-cpufreq-plugin xfce4-cpugraph-plugin xfce4-datetime-plugin \
```
so much tab

dnschneid commented on chroot-etc/intel-geteuid.patch:
```diff
@@ -0,0 +1,39 @@
+ 
+ 	/* Non-root user holding MASTER, don't let go */
+-	if (geteuid() && is_master(fd))
++	if (getuid0() && is_master(fd))
```
crouton invented the function getuid0 (which just returns 0), just because it's the same length as "geteuid" and is thus easy to binary hack.  But really, this patch should just convert geteuid() to the constant 0 and let the compiler optimize the rest out...

dnschneid commented on installer/gentoo/prepare:
```diff
@@ -0,0 +1,136 @@
+
+# install_pkg_dist: see install_pkg() in prepare.sh for details.
+install_pkg_dist() {
+    echo "GENTOO FIXME: install_pkg_dist"
```
I'm assuming these aren't actually going to be implemented?  In which case they need to error out more critically (as if you provided a bad parameter)

JoseJX replied:
Is erroring out with "exit 99" as you suggested before okay?

dnschneid replied:
Actually, it looks like [3 is the "correct" error code](https://github.com/dnschneid/crouton/blob/master/installer/functions#L134)

dnschneid reviewed (changes requested):
I'm running out of things to complain about! oh no!

JoseJX commented:
I tried last night to get this running on my old Samsung Chromebook 3, but I'm running into an issue where the unpacked stage tarball is bigger than /tmp. Do you have any suggestions for working around this?

I also need to look into the desktop switching bug that DennisLfromGA reported above, I can reproduce it on my Pixel 2.

Thanks again for reviewing!

dnschneid commented:
> I tried last night to get this running on my old Samsung Chromebook 3, but I'm running into an issue where the unpacked stage tarball is bigger than /tmp. Do you have any suggestions for working around this?

Whoa, how big is the stage3 tarball?  Maybe we'll need to add logic to main.sh to provide (and clean up) a temp folder on the same medium as the installation directory, if not doing `-d`.

JoseJX commented:
For ARM, it's about 1.1GB unpacked, plus the portage tree which is about 750MB. amd64/x86 should be around the same size. The files before unpacking are about 300MB for the stage tarball and around 90MB for the portage tree. We could probably save a lot of space by making some kind of sparse filesystem for the portage tree, the 4k blocksize my Pixel is using isn't really optimal for the huge number of small files in the portage tree.

Commit:
Let the user change the /tmp directory for low memory devices.

Commit:
Gentoo: Change xorg input targets to remove unused targets.

Commit:
Gentoo: Make Gentoo update chrome using the package manager.

Commit:
Gentoo: Add support for other arches, remove temporary files.

Commit:
Gentoo: Patch xf86-video-dummy and xf86-video-intel before the build.

JoseJX commented:
@DennisLfromGA - It looks like the lock isn't being removed on the switch between consoles, I'm still looking into it by working back from traces from a working Xenial install. Hopefully this will make clear what I'm doing wrong. Thanks for bringing it to my attention!

@dnschneid - I think most of the other issues are fixed now, but I wanted to bring up one thing. I have been thinking about the *-desktop targets and was thinking that instead of maintaining giant lists of packages to install, it might make more sense to just not allow those packages to install on Gentoo. So long as the base KDE/Gnome/lxde/xfce/whatever DE is installed, the user could pretty easily maintain the rest without really needing anything else from crouton.

So, since they're really sort of crutches for installing software that's maintained by the package manager and don't really come with anything extra (like startlxde or similar) would it be better to just not support those packages on Gentoo?

drinkcat commented on installer/gentoo/bootstrap:
```diff
@@ -0,0 +1,66 @@
+    echo 'ABI_X86="32 64"' >> "$DIR/etc/portage/make.conf"
+fi
+
+# Set beter cflags -- no, not -funroll-loops :)
```
nit: better

drinkcat commented on installer/main.sh:
```diff
@@ -80,6 +81,7 @@ Options:
                 Can only be specified alongside -m.
     -n NAME     Name of the chroot. Default is the release name.
                 Cannot contain any slash (/).
+    -o DIR	Override the temporary directory, use DIR instead of /tmp.
```
Nit: alignment.

drinkcat commented on targets/audio:
```diff
@@ -67,26 +71,32 @@ build_cras() {
+    if [ "${DISTROAKA:-"$DISTRO"}" = 'debian' ]; then
+        install --minimal --asdeps gcc $pkgdepextra libc6-dev$pkgsuffix \
+            pkg-config libspeexdsp-dev$pkgsuffix
+    fi
```
Prefix all the packages with debian= instead of using that if.

drinkcat commented on targets/chrome-common:
```diff
@@ -31,26 +31,52 @@ REQUIRES='x11'
+    keyword_package "$GENTOO_PKG" "$pkg"
+
+    # Accept the license
+    accept_license "$GENTOO_PKG" google-chrome "$pkg"
```
I'm mildly uncomfortable with this. How does Gentoo deal with this?

On Chromium OS, build_packages has an --accept-license flag.

JoseJX replied:
On Gentoo, you accept licenses by adding the license that you've accepted into the /etc/portage/package.licenses file, as it's done by the accept_license function. Alternatively, you'd use the --autounmask-write capability of emerge to write out these changes and then update the file using etc-update or a similar configuration merging tool.

I suppose I could ask the user if they accept the license as part of the install sequence? Is it okay to have this interaction in the install process?

To be fair though, shouldn't that really happen on the other distros too?

drinkcat replied:
> I suppose I could ask the user if they accept the license as part of the install sequence? Is it okay to have this interaction in the install process?

That's one possibility yes. We do have some automated tests though, so you'd need to make sure these interactions can be automated.

> To be fair though, shouldn't that really happen on the other distros too? I have to admit, I haven't tried installing chrome using one of the existing distros, maybe this is part of the install process?

That's a good point, we don't do anything on other distros. I guess it's fine this way then (especially since you take care of listing the licenses at the end of the install process).

drinkcat commented on targets/core:
```diff
@@ -106,17 +106,63 @@ EOF
+if [ "${DISTROAKA:-"$DISTRO"}" = 'gentoo' ]; then
+    # If the sudo group already exists, we'll assume we already set it up
+    if ! grep -q ^sudo /etc/group; then
+        groupadd -g 28 sudo
```
I don't think it's necessary to use gid 28 specifically?

drinkcat commented on targets/core:
```diff
@@ -106,17 +106,63 @@ EOF
+
+# Gentoo doesn't have a sudo group, add it now and allow those users to use sudo
+# Debian uses GID 28 for sudo
+if [ "${DISTROAKA:-"$DISTRO"}" = 'gentoo' ]; then
```
Do you need this test? Since other distros have a sudo group, the test below will fail and the group won't be created?

Commit:
Gentoo: Simplify audio dependency code.

Commit:
Gentoo: Simplify creating the sudo group in core.

dnschneid commented:
>  I think most of the other issues are fixed now, but I wanted to bring up one thing. I have been thinking about the *-desktop targets and was thinking that instead of maintaining giant lists of packages to install, it might make more sense to just not allow those targets to install on Gentoo. So long as the base KDE/Gnome/lxde/xfce/whatever DE is installed, the user could pretty easily maintain the rest without really needing anything else from crouton.

> So, since they're really sort of crutches for installing software that's maintained by the package manager and don't really come with anything extra (like startlxde or similar) would it be better to just not support those packages on Gentoo?

Yeah, I'd be okay with that.  Just make the -desktop targets depend on the normal targets and make gentoo not add anything in particular.  No need to fail out.

dnschneid commented:
w.r.t. 0269a1824270d337199f9eeddba1241a5039c2c8, I was actually hoping it would be more automatic.  Something like

1. main.sh creates two temporary directories.  One is in /tmp, the other is in $CHROOTSRC/.crouton-tmp or something (that's the real path, not the $CHROOT shadow in /var; don't want to go through ecryptfs).  If main.sh sees .crouton-tmp, it deletes it.
2. Bootstrap can pick which /tmp to use.  Ideally it would see how much space is available in tmpfs, but blindly picking one on a per-distro basis is fine too.
3. When main.sh deletes the /tmp directory, it deletes .crouton-tmp as well.

dnschneid commented on installer/gentoo/bootstrap:
```diff
@@ -0,0 +1,66 @@
+# has access to all of the variables set by main.sh, namely $tmp (the temporary
+# directory), $INSTALLERDIR/$DISTRO, $RELEASE, $ARCH, and $MIRROR.
+
+# 64 bit ARM processors
```
stale comment

dnschneid commented on installer/gentoo/defaults:
```diff
@@ -0,0 +1,25 @@
+case "$ARCH" in
+amd64 | x86_64) ARCH="amd64";;
+x86 | i?86) ARCH="x86";;
+armv7a_hardfloat | armv7hl) ARCH="armv7a_hardfloat";;
```
Maybe throw armhf in there as well, for good measure?

dnschneid commented on installer/gentoo/defaults:
```diff
@@ -0,0 +1,25 @@
+amd64 | x86_64) ARCH="amd64";;
+x86 | i?86) ARCH="x86";;
+armv7a_hardfloat | armv7hl) ARCH="armv7a_hardfloat";;
+armv7a | armv7l) ARCH="armv7a";;
```
and armel here.  For silly people coming from other distros :)

dnschneid commented on installer/gentoo/getrelease.sh:
```diff
@@ -0,0 +1,37 @@
+    # Get the CHOST architecture, first part of the tuple
+    CHOST="$(sed -n -e 's/^CHOST="\(.*\)*"/\1/p' "${2%/}$PM" | cut -d- -f1)"
+
+    case "$CHOST" in
```
weird question, but can we not trust the value of CHOST to be correct for its own distro?  I guess the only one that doesn't quite match is i686.

On ubuntu, we set the arch to i386, so maybe it'd be better to go with the gentoo arch naming for 32-bit intel.

JoseJX replied:
I reworked the naming again, so now, the architecture name should be consistent with the Gentoo arch keyword. I also added a few more architecture options.

dnschneid commented on targets/core:
```diff
@@ -106,17 +106,61 @@ EOF
+if [ "${DISTROAKA:-"$DISTRO"}" = 'gentoo' ]; then
+    # Install gentoolkit for euse, so we can set use flags
+    install gentoolkit
+    if [ "$ARCH" = 'x86' ] || [ "$ARCH" = 'amd64' ]; then
```
this would update if x86 became i686.  Also, `] || [` can be replaced with `-o`

dnschneid commented on targets/core:
```diff
@@ -106,17 +106,61 @@ EOF
+        # Install cpuid2cpuflags so we can easily set reasonable CFLAGS
+        # NOTE: This tool isn't really supported on ARM yet, update this later
+        install cpuid2cpuflags
+        CPUFLAGS=$(cpuinfo2cpuflags-x86 | cut -d: -f2)
```
quote the output? (actually not sure if you're intentionally collapsing whitespace or something)

dnschneid commented on targets/xorg:
```diff
@@ -197,6 +226,19 @@ if [ -z "$inteldriver" -a -n "$backport" ] && release -eq precise; then
         xterm x11-common xinput
+elif [ "${DISTROAKA:-"$DISTRO"}" = 'gentoo' ]; then
+    # Add use expand flags for the drivers we're installing
+    if echo "$installvideodrivers" | grep -q "$inteldriver"; then
```
`if [ "${installvideodrivers#*"$inteldriver"}" != "$installvideodrivers" ]; then` should do the same thing

Commit:
Gentoo: Add support for extra arm targets, fix ARCH variable usage.

Commit:
Gentoo: Rework intel and fbdev video_cards use expand.

Commit:
Gentoo: Add note for users installing *-desktop targets.

JoseJX commented:
After some package keywording (I'll take care of it on the Gentoo side, but it's not quite ready yet) I was able to run LXDE on my Chromebook 3 on ARM. Interestingly, console switching seems to be working just fine on that machine, so the issue is probably related to the Intel driver.

So here's my current list of things left to do:

1. Figure out why the console switching isn't working on Intel
2. Gnome/KDE
3. Kodi/XBMC
4. .desktop file for croutonurlhandler

Commit:
Gentoo: Adjust CPU use expand code to be more reliable.

Commit:
Gentoo: Set flags earlier to prevent recompiles. Add i915, i965 to VIDEO_CARDS.

Commit:
Gentoo: Support switching profiles. Add support for kde and gnome targets.

Fix.

Commit:
Gentoo: Add kodi support and make the xbmc package work with kodi.

Commit:
Gentoo: Update copyright year for all of the files I've touched.

JoseJX commented:
As an update:

- Gnome builds okay, but I'm having some issues getting it to start up correctly. It seems that the gnome-session-wrapper isn't actually starting the gnome session, so it defaults to twm/xterms. Still working on that.
- KDE builds, but I haven't really tested it at all yet
- I've looked into the issue with console switching and it seems that the issue is that the lock isn't released when switching consoles when using Gentoo/Intel. I've compared xprops between the two setups and traced through the switching process and it all seems to be working fine, but after the first switch, any attempt to lock /tmp/crouton-lock/display (fd==3) gets locked up. If you remove the lock, you can switch back to other running X instances. I did try using mawk instead of gawk, but that didn't seem to help either. I'll keep digging into this, but if anything seems familiar, I'd appreciate any input you've got.

Commit:
Gentoo: Fix issue where disabled flags would cause prepare script to halt.

nbe1233 commented:
If you have time perhaps you could implement musl and uclibc gentoo versions
https://wiki.gentoo.org/wiki/Project:Hardened_uClibc
https://wiki.gentoo.org/wiki/Project:Hardened_musl

JoseJX commented:
Yeah, hardened targets, x32 support and funtoo might be other interesting targets too, but for now I'm going to keep working on getting what I have in a merge-able state. Thanks for looking at it though!

I haven't had much time this week to dig into this, and really won't be able to for the following week either. If anyone has any ideas on the terminal switching issue, I'd really appreciate the input. If not, I'll get back on this in a week or so. Thanks!

zenmetsu commented:
Not sure what the current status of this branch is, but I did find a few issues.
The first was that the ARCH case statement in the installer was bombing with bad substitution on my Samsung Chromebook2.  `uname -m` reports `arm7l` on this machine.  I just hard-coded it to get the installer rolling.

The second issue is more of a Gentoo thing.  During the emerge world to update the base, there was a perl conflict which broke the process.  The solution was to change the emerge line in targets/core to `emerge -NDauv --with-bdeps y --backtrack 100 @world`

dnschneid commented:
@JoseJX, sorry for the silence here.  Could you please update the list of what isn't working?

JoseJX commented:
Sorry, I got busy with other work and haven't had too much time to work on this. I've still got the following issues open:

1. ARM changes to get it working for zenmetsu. I'd probably want to test on more ARM machines before calling any changes I make "fixed"
2. Console switching on Intel still seems to be broken. I can successfully switch back to Crouton, but not back to Gentoo from Crouton.
3. I haven't tested Gnome at all since I last posted here, I need to figure out what's going on with that not starting properly.
4. Documentation!

I'm hoping to have more free time in August. If you have any suggestions about the console switching issue, that's really the only big bug left.

idvoretskyi commented:
@JoseJX does CLI-only version of Gentoo chroot work now?

JoseJX commented:
On Intel (Chromebook Pixel 2 LS) and my ancient Chromebook (Samsung Series 3), yes, it does. I've been using it near daily since I started this bug. The graphical side starts fine, but won't switch consoles back to the Gentoo X11 instance on Intel. I still haven't figured out why.

I need to look into bringing in the changes from @zenmetsu to fix some issues on ARM as well, but this is more limited by my ability to test.

idvoretskyi commented:
Can you please provide some brief instructions what should I do to run
Gentoo CLI via Crouton?

I'm also using Chromebook Pixel LS, and would be happy to run Gentoo easily
on it.

On Aug 10, 2017 8:59 PM, "Joseph Jezak" <notifications@github.com> wrote:

> On Intel (Chromebook Pixel 2 LS) and my ancient Chromebook (Samsung Series
> 3), yes, it does. I've been using it near daily since I started this bug.
> The graphical side starts fine, but won't switch consoles back to the
> Gentoo X11 instance on Intel. I still haven't figured out why.
>
> I need to look into bringing in the changes from @zenmetsu
> <https://github.com/zenmetsu> to fix some issues on ARM as well, but this
> is more limited by my ability to test.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/dnschneid/crouton/pull/3175#issuecomment-321627703>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHOu_Mn7pRGy7roA_qCeE2RbGNlRHqcks5sW0UPgaJpZM4Mm4s7>
> .
>

Jokerswild2412 commented:
Has gentoo ever received full support for crouton? What are the steps for installation?  Is it necessary to follow the gentoo handbook for installation?

Thanks in advance!

Jokerswild2412 commented:
as a followup, I've attempted to download crouton-master from https://github.com/JoseJX/crouton/archive/master.tar.gz and attempted to install as normal in crouton, but got the following errors as listed in the screenshot:

![image](https://user-images.githubusercontent.com/7145744/38840381-4b3ef778-41ad-11e8-9b49-ee7dd3b3af68.png)

Did I do something wrong?  If not, what should I do to fix this error?

JoseJX commented:
Sorry, I've been really busy with paying work and haven't had much free hacking time! I suspect that things have changed a bit and it will take some work to get this working again. I will have a look this weekend since someone is interested in it, but things never worked 100% with X.

DennisLfromGA commented:
@JoseJX,

The 'wget' error should be an easy fix, it has been deprecated in CrOS, they're now using 'curl' instead.
See: Get rid of wget dependency, and workaround awk sandbox #3514

-**DennisLfromGA**

Jokerswild2412 commented:
@JoseJX 

I'm excited!  I've been wanting to get back to that gentoo life for awhile, I look forward to it.

Jokerswild2412 commented:
@JoseJX were you able to take a look at things this weekend? I'm still very interested in this.

Jokerswild2412 commented:
@JoseJX I'm still monitoring this, how are things going?

JL2210 commented:
I have a development Chromebook that I could use for this. Do you need any help? I have a PR on @JoseJX 's crouton.

JoseJX commented:
Sorry, my Chromebook Pixel 2 died a while ago and I haven't bought a new one yet. @JL2210, if you want to take this and run with it, be my guest!

JL2210 commented:
@JoseJX Sure thing. From what you've done, it shouldn't take much.

EnderNightLord-ChromeBook commented:
I will start testing for you too... Get some things out of the way... I might need some time though. I will test kde in my spear time.

JL2210 commented:
Not exactly Crouton.. but this script bootstraps Gentoo Prefix on a Chromebook (may need some work, but so far so good): https://gist.github.com/JL2210/f51a3d5a7a57900983b706949eba9f74

mpmilano commented:
I've forked this initiative and mostly updated it to work with 2020 gentoo, but we're back to the original issues @JoseJX was running into ---the VT switch back from ChromeOS to the running Gentoo X server really doesn't want to work.  I don't have any idea how this mechanism actually works; does anyone lurking on the thread have any insights? 

Fork here: https://github.com/mpmilano/crouton

This has been tested literally once, and presumably will break for some people.  If we can figure out the X11 issues then I'd be happy to clean up the fork.

TheGamingCoderNoob commented:
why am i here
